File: 2024/08/03/124. Binary Tree Maximum Path Sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res=INT_MIN;
    int path_sum(TreeNode* root){
        if (root == nullptr) {
            return 0;
        }
        int left=max(path_sum(root->left),0);
        int right=max(path_sum(root->right),0);
        res=max(res,left+right+root->val);
        return root->val+max(left,right);

    }
    int maxPathSum(TreeNode* root) {
        path_sum(root);
        return res;
    }
};

File: 2024/08/03/124. Binary Tree Maximum Path Sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        result=[root.val]

        def max_len(root):
            if root==None:
                return 0
            left_max=max(max_len(root.left),0)
            right_max=max(max_len(root.right),0)
            result[0]=max(result[0],root.val+left_max+right_max)
            return root.val+max(left_max,right_max)
        max_len(root)
        return result[0]


        
        